// fishfront/build/react/react-image.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

module.exports = React.createClass({
	displayName: 'exports',

	propTypes: {
		requireSrc: React.PropTypes.number,
		requireScale: React.PropTypes.number
	},
	updateImage: function updateImage(src) {
		var dom = this.refs.dom;
		var scale = this.props.scale;
		var width = dom.clientWidth;
		var height = dom.clientHeight;
		if (width != 0) {
			height = width * scale;
			dom.style.width = width + 'px';
			dom.style.height = height + 'px';
		} else {
			width = height / scale;
			dom.style.width = width + 'px';
			dom.style.height = height + 'px';
		}
		dom.setAttribute("src", src);
	},
	componentDidMount: function componentDidMount() {
		var self = this;
		setTimeout(function () {
			if (!self.isMounted()) {
				return;
			}
			self.updateImage(self.props.src);
		}, 0);
	},
	componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
		if (this.props.src != nextProps.src) this.updateImage(nextProps.src);
	},
	render: function render() {
		return React.createElement('img', _extends({}, this.props, { ref: 'dom' }));
	}
});
