// fishfront/build/util/uedit.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
function buildHtml(labelName, labelArgv) {
	var result = '<' + labelName;
	for (var i in labelArgv) {
		result += ' ' + i + '="' + labelArgv[i] + '"';
	}
	result += '></' + labelName + '>';
	return result;
}
function buildStyle(styleArgv) {
	var result = '';
	for (var i in styleArgv) {
		result += ' ' + i + ':' + styleArgv[i] + ';';
	}
	return result;
}
function replaceQQVideo2(content, articleWith) {
	var qqVideoRegex = /<embed.*?src="http(s?):\/\/v.qq.com\/.*?\/([^\/]*?)\.html.*?".*?width="(.*?)".*?height="(.*?)".*?\/>/img;
	var qqVideoWidth = articleWith;
	var qqVideoHeight = Math.ceil(qqVideoWidth / 1.5);
	var qqVideoSrc = 'https://v.qq.com/iframe/player.html?vid=$2&width=' + qqVideoWidth + '&height=' + qqVideoHeight + '&auto=0';
	var qqVideoStyle = buildStyle({
		'z-index': '1',
		'height': qqVideoHeight + 'px !important',
		'width': qqVideoWidth + 'px !important',
		'overflow': 'hidden'
	});
	var qqVideoReplace = buildHtml('iframe', {
		'class': 'video_iframe',
		'style': qqVideoStyle,
		'height': qqVideoHeight + 'px',
		'width': qqVideoWidth + 'px',
		'frameborder': '0',
		'data-src': qqVideoSrc,
		'src': qqVideoSrc,
		'allowfullscreen': '',
		'scrolling': 'no'
	});
	content = content.replace(qqVideoRegex, qqVideoReplace);
	return content;
}
function replaceQQVideo(content, articleWith) {
	var qqVideoRegex = /<embed.*?src="http(s?):\/\/static.video.qq.com\/TPout.swf\?vid=(.*?)".*?width="(.*?)".*?height="(.*?)".*?\/>/img;
	var qqVideoWidth = articleWith;
	var qqVideoHeight = Math.ceil(qqVideoWidth / 1.5);
	var qqVideoSrc = 'https://v.qq.com/iframe/player.html?vid=$2&width=' + qqVideoWidth + '&height=' + qqVideoHeight + '&auto=0';
	var qqVideoStyle = buildStyle({
		'z-index': '1',
		'height': qqVideoHeight + 'px !important',
		'width': qqVideoWidth + 'px !important',
		'overflow': 'hidden'
	});
	var qqVideoReplace = buildHtml('iframe', {
		'class': 'video_iframe',
		'style': qqVideoStyle,
		'height': qqVideoHeight + 'px',
		'width': qqVideoWidth + 'px',
		'frameborder': '0',
		'data-src': qqVideoSrc,
		'src': qqVideoSrc,
		'allowfullscreen': '',
		'scrolling': 'no'
	});
	content = content.replace(qqVideoRegex, qqVideoReplace);
	return content;
}
function replaceImage(content) {
	var imageRegex = /<img(.*?)\/?>/ig;
	var styleRegex = /style=\"(.*?)\"/ig;
	var imageReplace = '<img $1/>';
	var imageReplace2 = '<img $1 style="width: 100%"/>';
	content = content.replace(imageRegex, function (match, capture) {
		if (match.match(styleRegex)) {
			match = match.replace(imageRegex, imageReplace);
			match = match.replace(styleRegex, 'style="width:100%;$1"');
		} else {
			match = match.replace(imageRegex, imageReplace2);
		}
		return match;
	});
	return content;
}
exports.default = {
	parse: function parse(content, articleWith) {

		content = replaceQQVideo(content, articleWith);

		content = replaceQQVideo2(content, articleWith);

		content = replaceImage(content);

		return content;
	}
};
