// fishfront/src/react/react-document-head.js
import React from 'react';
import Env from '../runtime/env';

export default React.createClass({
	contextTypes:{
        documentHead:React.PropTypes.object,
    },
	getInitialState(){
		if( Env.isInBrowser() ){
			this.updateTitle(this.props.title);
		}else{
			this.updateContext( this.props );
		}
		return {};
	},
	componentWillReceiveProps(nextProps){
		if( Env.isInBrowser() ){
			if(nextProps.title != this.props.title)
				this.updateTitle( nextProps.title );
		}else{
			this.updateContext( nextProps );
		}
	},
	updateTitle(title){
		if( !title )
			return;
		var body = document.body;
		document.title = title;
		var iframe = document.createElement("iframe");
		iframe.setAttribute("key",this.props.title);
		iframe.setAttribute("style","display:none;width:0px;height:0px");
		iframe.setAttribute("src",'/logo.ico');
 
		body.appendChild(iframe);
		iframe.addEventListener('load',function() { 
			setTimeout(function() {   
			    iframe.removeEventListener('load',function(){}); 
			    body.removeChild(iframe);
			}, 0);  
		});
	},
	updateContext(nextProps){
		if( this.context && this.context.documentHead ){
			this.context.documentHead.setHeader(nextProps);
		}
	},
	render(){
		return null;
	}
});
