// fishfront/src/react/react-image.js
module.exports =  React.createClass({
	propTypes: {
		requireSrc      : React.PropTypes.number,
		requireScale    : React.PropTypes.number,
    },
    updateImage:function(src){
    	var img = new Image();
    	var dom = this.refs.dom;
		var scale = this.props.scale;
		var width = dom.clientWidth;
		var height = dom.clientHeight;
		dom.style.background = 'grey';
		if( width != 0 ){
			height = width*scale;
			dom.style.width = width+'px';
			dom.style.height = height+'px';
		}else{
			width = height/scale;
			dom.style.width = width+'px';
			dom.style.height = height+'px';
		}
		img.onload = function(){
			dom.style.background = 'none';
		};
		img.src = src;
		dom.setAttribute("src",src);
    },
	componentDidMount: function(){
		var self = this;
		setTimeout(function(){
			if( !self.isMounted() ){
				return;
			}
			self.updateImage( self.props.src );
		},0);
	},
	componentWillReceiveProps:function(nextProps){
		if( this.props.src != nextProps.src )
			this.updateImage( nextProps.src );
	},
	render: function() {
		return <img {...this.props} ref="dom"/>;
	}
});
